Source/Core/Common/BlockingLoop.h

```diff
@@ -132,64 +133,46 @@ class BlockingLoop
 
     while (!m_shutdown.IsSet())
     {
-      payload();
+      if (m_running_state.load() == RunningState::NeedExecution)
+      {
+        m_running_state.store(RunningState::Executing);
+
+        payload();
 
-      switch (m_running_state.load())
+        // If we're still in the Running state, then Wakeup wasn't called within the last execution
+        // of the payload. This means we should be ready now. Else we're in the Done state now, so
+        // wakeup the waiting threads right now.
+        RunningState expected_running{RunningState::Executing};
+        if (m_running_state.compare_exchange_strong(expected_running, RunningState::Done))
+          m_done_event.Set();
+      }
+      // We're done now. So time to check if we want to sleep or if we want to stay in a busy loop.
+      else if (m_may_sleep.TestAndClear())
       {
-      case STATE_NEED_EXECUTION:
-        // We won't get notified while we are in the STATE_NEED_EXECUTION state, so maybe Wakeup was
-        // called.
-        // So we have to assume on finishing the STATE_NEED_EXECUTION state, that there may be some
-        // remaining tasks.
-        // To process this tasks, we call the payload again within the STATE_LAST_EXECUTION state.
-        m_running_state--;
-        break;
-
-      case STATE_LAST_EXECUTION:
-        // If we're still in the STATE_LAST_EXECUTION state, then Wakeup wasn't called within the
-        // last
-        // execution of the payload. This means we should be ready now.
-        // But bad luck, Wakeup may have been called right now. So break and rerun the payload
-        // if the state was touched.
-        if (m_running_state-- != STATE_LAST_EXECUTION)
-          break;
-
-        // Else we're likely in the STATE_DONE state now, so wakeup the waiting threads right now.
-        // However, if we're not in the STATE_DONE state any more, the event should also be
-        // triggered so that we'll skip the next waiting call quite fast.
-        m_done_event.Set();
-
-      case STATE_DONE:
-        // We're done now. So time to check if we want to sleep or if we want to stay in a busy
-        // loop.
-        if (m_may_sleep.TestAndClear())
-        {
-          // Try to set the sleeping state.
-          if (m_running_state-- != STATE_DONE)
-            break;
-        }
-        else
-        {
-          // Busy loop.
-          break;
-        }
+        // Try to set the sleeping state.
+        RunningState expected_done{RunningState::Done};
+        if (!m_running_state.compare_exchange_strong(expected_done, RunningState::Sleeping))
```

Comment:
Isn't compare_exchange_weak fine here as well?

Reply:
Hmm. Yeah `m_may_sleep` would have to be reset.. I'm not sure how to distinguish spurious failures, when we'd want to reset it, vs. actual failures, when the user presumably wants it to go back into the busy loop.

On my system, making this `compare_exchange_weak` doesn't affect the time to run BlockingLoopTest.